Changes in Pixel 2D template making code cfg for Phase-2 + adding a unit test

tvami opened the pull request:
PR description:

Minor modifications to the Pixel 2D template making code for the Phase-2 detector. After #31622 we are able to make charge reweighting in Phase-2, too, which will need the 2D templates.  Also added a unit test which creates a 2D template file.

PR validation:
The command which is in the unit test was run.

if this PR is a backport please specify the original PR and why you need to backport that PR:
This is not a PR, and there is no PR planned

cc: @mmusich

Commit:
Changes in 2D template making code cfg + adding a unit test

tvami commented:
Please test it together with 
https://github.com/cms-data/CalibTracker-SiPixelESProducers/pull/2

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31796/19076

- This PR adds an extra 20KB to repository

cmsbuild commented:
A new Pull Request was created by  @tvami (Tamas Vami) for master.

It involves the following packages:

CondTools/SiPixel


@ggovi, @cmsbuild can you please review it and eventually sign? Thanks.
@ferencek, @mmusich, @dkotlins, @VinInn, @tvami this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mmusich commented:
test parameters
* pull_request =  cms-data/CalibTracker-SiPixelESProducers#2

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/CalibTracker-SiPixelESProducers#2

cmsbuild commented:
-1

Tested at: 7d85b0025a4997a00176a6dc77382152ac9795c2

CMSSW: CMSSW_11_2_X_2020-10-14-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/9969/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  HeaderConsistency  ClangBuild**

* **Build**:

I found compilation error when building: 

<pre>Entering library rule at HeterogeneousCore/SonicTriton
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/src/TritonClient.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/src/triton_utils.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/src/TritonData.cc
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/src/triton_utils.cc:1:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/interface/triton_utils.h:9:10: fatal error: request_grpc.h: No such file or directory
 #include "request_grpc.h"
          ^~~~~~~~~~~~~~~~
compilation terminated.
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/src/TritonData.cc:1:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/interface/TritonData.h:15:10: fatal error: request_grpc.h: No such file or directory
</pre>

* **Clang**:

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>Entering library rule at HeterogeneousCore/SonicTriton
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/src/TritonClient.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/src/TritonData.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/src/triton_utils.cc
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/src/triton_utils.cc:1:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/interface/triton_utils.h:9:10: fatal error: 'request_grpc.h' file not found
#include "request_grpc.h"
         ^~~~~~~~~~~~~~~~
1 error generated.
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/src/TritonData.cc:1:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/HeterogeneousCore/SonicTriton/interface/TritonData.h:15:10: fatal error: 'request_grpc.h' file not found
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

mmusich commented:
let's ty again, hopefully with a new IB.

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/CalibTracker-SiPixelESProducers#2

cmsbuild commented:
-1

Tested at: 7d85b0025a4997a00176a6dc77382152ac9795c2

CMSSW: CMSSW_11_2_X_2020-10-15-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/9990/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testPixelDBs had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/9990/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2543752
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2543723
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

mmusich commented on CondTools/SiPixel/test/BuildFile.xml:
```diff
@@ -19,3 +19,9 @@
 </environment>
 
+<environment>
+  <bin   file="testPixelDBs.cpp">
```
@tvami this new test used the same `testPixelDBs.cpp` file as the already existing test `run_0T_BoR3`.
Either split it (creating a new `testPixePhaselDBs.cpp`) or put everything in the same bash executable.

mmusich commented on CondTools/SiPixel/test/run_Phase2cpeDBTest_v6.sh:
```diff
@@ -1,4 +1,10 @@
-cmsRun SiPixelTemplateDBObjectUploader_Phase2_cfg.py MagField=3.8 Version=6 Append=mc_25x100 Map=../data/phase2_T15_mapping.csv geometry=T15 TemplateFilePath=CondTools/SiPixel/data/25by100_dual_slope_temps
```
@tvami this file is missing the shebang.

mmusich commented on CondTools/SiPixel/test/run_Phase2cpeDBTest_v6.sh:
```diff
@@ -1,4 +1,10 @@
+cmsRun  ${LOCAL_TEST_DIR}/SiPixelGenErrorDBObjectUploader_Phase2_cfg.py MagField=3.8 Version=6 Append=mc_25x100 Map= ${LOCAL_TEST_DIR}/../data/phase2_T15_mapping.csv geometry=T15 GenErrFilePath= ${LOCAL_TEST_DIR}/CondTools/SiPixel/data/25by100_dual_slope_temps || die "Failure running SiPixelGenErrorDBObjectUploader_Phase2_cfg.py" $? 
+
+#For 2D templates 
+cmsRun  ${LOCAL_TEST_DIR}/SiPixel2DTemplateDBObjectUploader_Phase2_cfg.py MagField=3.8 Version=7 Append=mc_25x100 Map= ${LOCAL_TEST_DIR}/../data/Phase2_25by100_templates_2020October/phase2_T15_mapping.csv TemplateFilePath= ${LOCAL_TEST_DIR}/CondTools/SiPixel/data/Phase2_25by100_templates_2020October denominator=True || die "Failure running SiPixel2DTemplateDBObjectUploader_Phase2_cfg.py" $? 
```
while you are at it, please fix the missing newline.

Commit:
Putting the unit tests into one file

Putting the unit tests into one file

Minor fix

Minor fix

Minor

Alternative way to reach the mapping file

fix

fix3

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31796/19146

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #31796 was updated. @ggovi, @cmsbuild can you please check and sign again.

mmusich commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/CalibTracker-SiPixelESProducers#2

cmsbuild commented:
-1

Tested at: 7fed1623ad9bfe6cf1d2cf7e0a53886cca26862a

CMSSW: CMSSW_11_2_X_2020-10-16-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/10032/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testPixelDBs had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31796/19158

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #31796 was updated. @ggovi, @cmsbuild can you please check and sign again.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/10032/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2543752
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2543723
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/CalibTracker-SiPixelESProducers#2

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31796/19208

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #31796 was updated. @ggovi, @cmsbuild can you please check and sign again.

mmusich commented:
@cmsbuild, please abort

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/CalibTracker-SiPixelESProducers#2

cmsbuild commented:
-1

Tested at: 0c37cdf2362b79dba24f6adb711950a9a2306c65

CMSSW: CMSSW_11_2_X_2020-10-19-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/10092/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testPixelDBs had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/10092/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2543752
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2543729
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31796/19224

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #31796 was updated. @ggovi, @cmsbuild can you please check and sign again.

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/CalibTracker-SiPixelESProducers#2

cmsbuild commented:
-1

Tested at: d0d71bf50d94278e6be3cabc50d671a7678d3764

CMSSW: CMSSW_11_2_X_2020-10-19-2300
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/10127/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testPixelDBs had ERRORS

cmsbuild commented:
Comparison job queued.

mrodozov commented on CondTools/SiPixel/test/createTestDBObjects.sh:
```diff
@@ -11,10 +11,10 @@ echo "TESTING Pixel 1D GenErr DB code ..."
 
 echo "TESTING Pixel 1D Template DB code for Phase-2 ..."
-cmsRun ${LOCAL_TEST_DIR}/SiPixelTemplateDBObjectUploader_Phase2_cfg.py MagField=3.8 Version=7 Append=mc_25x100 Map=${LOCAL_TEST_DIR}/../data/Phase2_25by100_templates_2020October/phase2_T15_mapping.csv geometry=T15 TemplateFilePath=CondTools/SiPixel/data/Phase2_25by100_templates_2020October || die "Failure running SiPixelTemplateDBObjectUploader_Phase2_cfg.py" $? 
+cmsRun ${LOCAL_TEST_DIR}/SiPixelTemplateDBObjectUploader_Phase2_cfg.py MagField=3.8 Version=7 Append=mc_25x100 Map=$CMSSW_BASE/src/CondTools/SiPixel/data/Phase2_25by100_templates_2020October/phase2_T15_mapping.csv geometry=T15 TemplateFilePath=CondTools/SiPixel/data/Phase2_25by100_templates_2020October || die "Failure running SiPixelTemplateDBObjectUploader_Phase2_cfg.py" $? 
```
I'm not sure CalibTracker-SiPixelESProducers and CondTools/SiPixel are the same thing. Would you try chaning this to CalibTracker/SiPixelESProducers becase your PR is actually for that package/subpackage

cmsbuild commented:
The code-checks are being triggered in jenkins.

tvami commented:
Hi @mrodozov , I did your suggestion, thanks! Can you please trigger tests?

mrodozov commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31796/19260

- This PR adds an extra 16KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/CalibTracker-SiPixelESProducers#2

cmsbuild commented:
Pull request #31796 was updated. @ggovi, @cmsbuild can you please check and sign again.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/10127/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544100
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544071
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

cmsbuild commented:
-1

Tested at: 3206e1a9ecf05cc1e0400297fc504bdb5a92ef17

CMSSW: CMSSW_11_2_X_2020-10-20-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/10153/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testPixelDBs had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31796/19273

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #31796 was updated. @ggovi, @cmsbuild can you please check and sign again.

mrodozov commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/CalibTracker-SiPixelESProducers#2

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/10153/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544100
 - DQMHistoTests: Total failures: 10
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544068
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

cmsbuild commented:
+1
Tested at: c9b5e8ee7bf1440889e7028b2ff233c8d80fb492
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/10161/summary.html
CMSSW: CMSSW_11_2_X_2020-10-20-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c7173/10161/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544100
 - DQMHistoTests: Total failures: 10
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544068
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

ggovi commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1

mrodozov commented:
After this PR this test:
SiPixelTemplateDBObjectUploader_Phase2_cfg
started failing for the python3 IBs (PY3 and DEVEL). 
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc820/CMSSW_11_2_PY3_X_2020-11-06-2300/unitTestLogs/CondTools/SiPixel#/18645-18645
There are two things incompatible with Py3  - the print statements and the different csv module interface.
The print is easy, but the cvs statements might need be put in if py2 elseif py3 statements.
